File: main/unittransfer/health.py
"""One door to every check (Phase 54, M17).

The toolkit has more validators than any of the reference tools and, until
this, no single place that ran them: the map rules live on the map screen, the
faction audit on the faction screen, the EDB rules on the buildings screen and
the file checkers inside each file's editor. Somebody whose mod crashes had to
know all of that before they could ask "what is wrong with it".

This module asks every one of them and hands back one list, in one shape. It
owns no rule. Each **source** is an adapter from one validator's own answer to
a :class:`Finding`, and the screen that owns a finding is still the place to
fix it - every finding says which screen that is.

**Only the fast ones run.** Measured cold on the two installed mods on
2026-09-22: the map rules 0.9s and 2.8s, the faction audit about 1.2s, the EDB
checks 0.2s. The four cleanup audits cost 7s to 95s each (cards on ROCSS is the
95) and are about a tidy mod rather than one that starts, so they are listed
in :data:`SLOW` as a door into their own screens and never run from here.

**A source that throws is a failed source**, reported with its error, never an
empty one and never the whole answer. That is `mapcheck.run`'s rule for its own
rules, one level up: losing the faction audit to a broken trait file would be
the failure that makes a dashboard untrustworthy.

**Severity.** The map and EDB rules already say `fatal`/`warn`/`note`. Several
file checkers say `fatal` as a bool, and three (traits, ancillaries, factions)
say nothing at all - but their sentences do: "more than 8 crashes the game",
"the game stops loading the file here". :func:`severity_of` reads that, so a
checker that learns a new crash is ranked by the sentence it already writes,
with no second table here to forget to update.

**When** is the column the two crash guides add. Nobody arrives saying "my
EDB has a fatal"; they arrive saying "it crashes loading the campaign". Each
source says when its findings usually bite, and a sentence that names a panel
or a battle moves its finding there.
"""
from __future__ import annotations

import re
import time
from dataclasses import dataclass, field, asdict
from pathlib import Path
from typing import Callable, Dict, Iterable, List, Optional

SEVERITIES = ("fatal", "warn", "note")

#: When a finding bites, in the order somebody playing meets them.
WHEN = (
    ("launch", "Starting the game", "Read before the main menu: the unit, trait, "
     "ancillary and faction files. A fault here is a crash or a black screen at start."),
    ("campaign", "Loading a campaign", "The map, the settlements, the buildings and "
     "who starts where. A fault here crashes the campaign on its loading screen."),
    ("panel", "Opening a panel", "Only when a particular scroll, detail screen or "
     "building browser is opened."),
    ("battle", "In battle", "When a battle loads the models and the units in it."),
    ("play", "During play", "Later: an AI turn, an event firing, a character "
     "gaining a trait."),
)
WHEN_IDS = tuple(w[0] for w in WHEN)

# A checker's own sentence, read for how bad it is. Kept to phrases the checkers
# actually write (see tests/test_health.py, which holds these to real messages).
_FATAL_WORDS = re.compile(
    r"crash|stops loading|will not load|won't load|does not load|stops recognising|"
    r"refuses to load|can(?:no|')t be read|cannot start|black screen|ctd\b",
    re.I)
_PANEL_WORDS = re.compile(r"detail screen|scroll|panel|building browser|"
                          r"when (?:it is )?opened|opening", re.I)
_BATTLE_WORDS = re.compile(r"\bbattle\b", re.I)


def severity_of(raw: dict, default: str = "warn") -> str:
    """``fatal``/``warn``/``note`` for one checker's finding.

    An explicit ``severity`` wins; then ``fatal: True``; then the sentence."""
    sev = str(raw.get("severity") or "").lower()
    if sev in SEVERITIES:
        return sev
    if raw.get("fatal") is True:
        return "fatal"
    if _FATAL_WORDS.search(str(raw.get("message") or "")):
        return "fatal"
    return default


def when_of(message: str, default: str) -> str:
    if _PANEL_WORDS.search(message or ""):
        return "panel"
    if _BATTLE_WORDS.search(message or ""):
        return "battle"
    return default


@dataclass
class Finding:
    """One thing wrong, whichever validator found it.

    ``what`` is the identity and, as in :class:`mapcheck.Finding`, never holds a
    line number. ``open`` is where the page goes to show it: a mode, and the
    record in that mode when there is one."""

    source: str
    code: str
    severity: str
    message: str
    file: str = ""
    line: int = 0
    what: str = ""
    when: str = "campaign"
    count: int = 1
    #: already there when the mod's baseline was taken (map rules only)
    baseline: bool = False
    open: Dict[str, str] = field(default_factory=dict)


@dataclass
class Source:
    id: str
    label: str
    #: the mode the page opens for this source's findings
    mode: str
    file: str
    when: str
    run: Callable[..., List[Finding]]
    #: needs the campaign map, so off where the map screen is off
    needs_map: bool = False


SOURCES: List[Source] = []


def source(id: str, label: str, mode: str, file: str, when: str, needs_map: bool = False):
    def wrap(fn):
        SOURCES.append(Source(id, label, mode, file, when, fn, needs_map))
        return fn
    return wrap


def _plain(rows: Iterable[dict], src: str, file: str, when: str, mode: str,
           name_key: str, code_prefix: str) -> List[Finding]:
    """The file checkers' shape: ``{kind|code, <name>, line?, message, fatal?}``."""
    out = []
    for r in rows:
        msg = str(r.get("message") or "")
        name = str(r.get(name_key) or r.get("name") or "")
        code = str(r.get("code") or r.get("kind") or "finding")
        out.append(Finding(
            source=src, code=f"{code_prefix}.{code}", severity=severity_of(r),
            message=(f"{name}: {msg}" if name and name not in msg else msg),
            file=file, line=int(r.get("line") or 0),
            what=f"{name}|{code}", when=when_of(msg, when),
            open={"mode": mode, "name": name}))
    return out


# ---------------------------------------------------------------------------
# the sources


@source("map", "Campaign map rules", "campmap", "world/maps", "campaign", needs_map=True)
def _map(mod, ctx) -> List[Finding]:
    from . import mapcheck
    cm = ctx["map_for"](ctx.get("campaign") or "")
    rep = mapcheck.run(mod, cm, ctx.get("campaign") or "")
    out = []
    for f in rep.findings:
        out.append(Finding(
            source="map", code=f.code, severity=f.severity, message=f.message,
            file=f.file, line=f.line, what=f.what, count=f.count,
            baseline=f.baseline, when=when_of(f.message, "campaign"),
            open={"mode": "campmap", "key": f.key}))
    for r in rep.failed:
        out.append(Finding(
            source="map", code=str(r.get("code") or "rule"), severity="warn",
            message=f"this rule could not run: {r.get('error') or r}",
            what=f"failed|{r.get('code')}", when="campaign",
            open={"mode": "campmap"}))
    return out


@source("factions_audit", "Is every faction complete", "factions",
        "descr_sm_factions.txt and five more", "campaign")
def _faction_audit(mod, ctx) -> List[Finding]:
    from . import factionaudit
    rep = factionaudit.audit(mod, ctx.get("campaign") or "")
    out = []
    if rep.get("error"):
        out.append(Finding("factions_audit", "audit.roster", "fatal", rep["error"],
                           file="descr_sm_factions.txt", what="roster",
                           when="launch", open={"mode": "factions"}))
    for fac in rep.get("factions") or []:
        for row in fac.get("rows") or []:
            if row.get("state") != "missing":
                continue
            sev = "warn" if row.get("level") == "gap" else "note"
            out.append(Finding(
                source="factions_audit", code=f"audit.{row.get('id')}", severity=sev,
                message=f"{fac.get('label') or fac.get('slot')}: {row.get('label')} - "
                        f"{row.get('detail') or 'missing'}",
                file=str(row.get("rel") or ""), what=f"{fac.get('slot')}|{row.get('id')}",
                when="campaign", open={"mode": "factions", "name": str(fac.get("slot") or "")}))
    return out


@source("edb", "Buildings: the tree and recruitment", "buildings",
        "export_descr_buildings.txt", "campaign")
def _edb(mod, ctx) -> List[Finding]:
    from . import buildings
    rep = buildings.checks(mod)
    out = []
    for f in (rep.get("tree") or {}).get("findings") or []:
        out.append(Finding(
            source="edb", code=f["code"], severity=f["severity"], message=f["message"],
            file=buildings.EDB_REL, line=int(f.get("line") or 0), what=f.get("what", ""),
            count=int(f.get("count") or 1), when=when_of(f["message"], "campaign"),
            open={"mode": "buildings", "name": f.get("building", "")}))
    # The recruitment checks are the unit editor's half: a unit that stops being
    # trainable up a chain, or trained twice by one level. Worth knowing, never a
    # crash, so notes - and folded to one per building line, because DaC has
    # 2 151 of them unit by unit and a door that opens on two thousand notes is
    # a door nobody reads past. The building's own Checks panel lists each one.
    # The city/castle comparison is a tool, not a finding, and is left out.
    for ln in rep.get("lines") or []:
        for key, code, says in (
                ("gaps", "recruit.gap", "not trained at every level above the one that first trains them"),
                ("dupes", "recruit.twice", "trained twice by one level")):
            units = sorted({str(x.get("unit") or "") for x in (ln.get(key) or [])})
            if not units:
                continue
            shown = ", ".join(units[:4]) + (f" and {len(units) - 4} more" if len(units) > 4 else "")
            out.append(Finding(
                source="edb", code=code, severity="note", count=len(units),
                message=f"{ln['line']}: {len(units)} unit{'s' if len(units) != 1 else ''} "
                        f"{says} ({shown})",
                file=buildings.EDB_REL, what=f"{ln['line']}|{code}",
                when="play", open={"mode": "buildings", "name": ln["line"]}))
    return out


@source("edu", "Units: the engine's ceilings", "edit", "export_descr_unit.txt", "launch")
def _edu(mod, ctx) -> List[Finding]:
    from . import educeil, modflags
    rows = modflags.uncapped(educeil.mod_findings(mod.edu.units), mod)
    return _plain(rows, "edu", "export_descr_unit.txt", "launch", "edit", "name", "edu")


@source("traits", "Traits", "traits", "export_descr_character_traits.txt", "launch")
def _traits(mod, ctx) -> List[Finding]:
    from . import traits, triggers, modflags
    path = Path(mod.edct_path)
    if not path.exists():
        return []
    rows = modflags.uncapped(traits.check_file(traits.parse_file(path),
                                               triggers.parse_file(path)), mod)
    return _plain(rows, "traits", path.name, "launch", "traits", "trait", "trait")


@source("ancillaries", "Ancillaries", "ancillaries", "export_descr_ancillaries.txt", "launch")
def _ancillaries(mod, ctx) -> List[Finding]:
    from . import ancillaries, triggers, modflags
    path = Path(mod.eda_path)
    if not path.exists():
        return []
    rows = modflags.uncapped(ancillaries.check_file(ancillaries.parse_file(path),
                                                    triggers.parse_file(path), mod), mod)
    return _plain(rows, "ancillaries", path.name, "launch", "ancillaries",
                  "ancillary", "anc")


@source("factions", "Factions", "factions", "descr_sm_factions.txt", "launch")
def _factions(mod, ctx) -> List[Finding]:
    from . import factions, modflags
    path = factions.path_for(mod)
    if not path.is_file():
        return []
    rows = modflags.uncapped(factions.check_file(factions.parse_file(path), mod), mod)
    return _plain(rows, "factions", factions.REL, "launch", "factions", "name", "faction")


@source("guilds", "Guilds", "guilds", "export_descr_guilds.txt", "play")
def _guilds(mod, ctx) -> List[Finding]:
    from . import guilds
    if not guilds.path_for(mod).exists():
        return []
    gf, text = guilds.read(mod)
    rows = guilds.check_file(gf, guilds.trigger_file(text), guilds.building_names(mod))
    return _plain(rows, "guilds", guilds.GUILDS_REL, "play", "guilds", "guild", "guild")


@source("campdb", "Campaign constants", "campdb", "descr_campaign_db.xml", "campaign")
def _campdb(mod, ctx) -> List[Finding]:
    from . import campdb
    if not campdb.path_for(mod).exists():
        return []
    db, _ = campdb.read(mod)
    # "key" is what a campdb finding calls its tag ("agents/assassinate_chance_max"),
    # and it is what the screen's cdbOpen takes - `name` would come back empty and
    # the row would land on the screen with nothing picked.
    return _plain(campdb.check_file(db), "campdb", campdb.REL, "campaign", "campdb",
                  "key", "campdb")


@source("minor", "Rebels, religions, resources, cultures, names", "minor",
        "five campaign files", "campaign")
def _minor(mod, ctx) -> List[Finding]:
    from . import minorfiles
    out = []
    for t in minorfiles.TABS:
        if not minorfiles.path_for(mod, t.id).is_file():
            continue
        parsed, _ = minorfiles.read_any(mod, t.id)
        rows = minorfiles.check_any(mod, t.id, parsed)
        mode = "cultures" if t.id == "cultures" else "minor"
        for f in _plain(rows, "minor", t.rel, "campaign", mode, "name", t.id):
            if mode == "minor":
                f.open["tab"] = t.id
            out.append(f)
    return out


@source("crash", "What the crash guides name", "rawtext", "five files", "play")
def _crash(mod, ctx) -> List[Finding]:
    """Phase 54b's rules: the guides' causes that belonged to no module."""
    from . import crashrules
    found, failed = crashrules.run(mod)
    out = []
    for r, f in found:
        open_ = {"mode": r.mode}
        if r.mode == "rawtext":
            open_.update(rel=f.file, line=str(f.line))
        elif f.name:
            open_["name"] = f.name
        out.append(Finding(
            source="crash", code=f.code, severity=f.severity, message=f.message,
            file=f.file, line=f.line, what=f.what, when=f.when, open=open_))
    for x in failed:
        out.append(Finding(
            source="crash", code=x["code"], severity="warn",
            message=f"this rule could not run: {x['error']}",
            what=f"failed|{x['code']}", when="play", open={"mode": "rawtext"}))
    return out


#: The cleanup audits: listed, never run here (see the module docstring).
SLOW = (
    {"id": "bmdb", "label": "Battle models no unit uses, and files nothing names",
     "mode": "bmdb", "cost": "10 to 30 seconds"},
    {"id": "dupes", "label": "Battle models entered twice (the game reads the first)",
     "mode": "bmdb", "cost": "7 to 20 seconds"},
    {"id": "stratmap", "label": "Strat map models nothing draws",
     "mode": "stratmap", "cost": "7 to 16 seconds"},
    {"id": "cards", "label": "Unit cards for units that are gone, and copies",
     "mode": "cards", "cost": "25 to 95 seconds"},
)


def _refused() -> List[dict]:
    from . import crashrules
    return [dict(x) for x in crashrules.REFUSED]


# ---------------------------------------------------------------------------
# the run


def run(mod, map_for: Optional[Callable] = None, campaign: str = "",
        only: Iterable[str] = ()) -> dict:
    """Every fast source over one mod. ``map_for(campaign)`` gives the map, and
    without it the map source is skipped (the build has the map screen off)."""
    ctx = {"map_for": map_for, "campaign": campaign}
    want = set(only or ())
    findings: List[Finding] = []
    sources = []
    t_all = time.perf_counter()
    for s in SOURCES:
        if want and s.id not in want:
            continue
        row = {"id": s.id, "label": s.label, "mode": s.mode, "file": s.file,
               "when": s.when, "state": "ok", "error": "", "ms": 0,
               "counts": {k: 0 for k in SEVERITIES}}
        if s.needs_map and map_for is None:
            row["state"] = "off"
            sources.append(row)
            continue
        t = time.perf_counter()
        try:
            got = s.run(mod, ctx)
        except Exception as e:                      # noqa: BLE001 - the point
            row["state"] = "failed"
            row["error"] = f"{type(e).__name__}: {e}"
            got = []
        row["ms"] = int((time.perf_counter() - t) * 1000)
        for f in got:
            if f.severity not in SEVERITIES:
                f.severity = "warn"
            if f.when not in WHEN_IDS:
                f.when = s.when
            row["counts"][f.severity] += 1
        findings += got
        sources.append(row)
    order = {s: i for i, s in enumerate(SEVERITIES)}
    findings.sort(key=lambda f: (order[f.severity], f.baseline, WHEN_IDS.index(f.when),
                                 f.source, f.message))
    return {
        "mod": getattr(mod, "name", ""), "campaign": campaign,
        "ms": int((time.perf_counter() - t_all) * 1000),
        "findings": [asdict(f) for f in findings],
        "counts": {k: sum(1 for f in findings if f.severity == k) for k in SEVERITIES},
        "sources": sources,
        "slow": [dict(x) for x in SLOW],
        # what the crash guides claim and measuring refused, so it is findable
        "refused": _refused(),
        "when": [{"id": w[0], "label": w[1], "help": w[2]} for w in WHEN],
    }

File: main/web/js/health.js
/* health.js - Health: every check the toolkit has, over one mod (Phase 54, M17)

   Part of the Medieval 2 GUI Toolkit UI. These files are plain
   <script> tags sharing ONE global scope, loaded in the order set in
   index.html - there is no build step and no module system. Two rules
   follow from that: a top-level name must be unique across all of
   them, and a file's top-level side effects may not depend on a file
   loaded after it. */
/* =====================================================================
   A DOOR, NOT A SECOND COPY. The server (unittransfer/health.py) asks each
   validator for its findings and hands back one list in one shape; this page
   decides nothing about any of them. A row says what is wrong, which file,
   and opens the screen that owns it, which is where it is fixed.

   Grouped by WHEN it bites, because that is how somebody arrives: "it crashes
   loading the campaign", not "my EDB has a fatal". Fatal first inside each.

   Notes start hidden: a shipping mod carries a thousand of them (DaC's
   recruitment notes alone fold down to ninety), and a page that opens on those
   is a page nobody reads to the fatal at the top of.

   Every name here starts `hl`; there was none in the tree before.
   ===================================================================== */

const HL_SEV = {
  fatal: ['✕', 'w-bad', 'Fatal: the game crashes or refuses the file'],
  warn: ['!', 'w-warn', 'Warning: it loads, but something will not work as meant'],
  note: ['·', '', 'Note: worth a look, not a fault'],
};
const HL_PAGE = 150;

function hlNew(mod){
  return {mod, rep: null, err: '', busy: false, notes: false, base: false,
          src: '', shown: {}};
}

function renderHealth(){
  const h = state.hl;
  if(!h || h.mod !== state.src){ loadHealth(); return; }
  hlPaint();
}

async function loadHealth(){
  const mod = state.src;
  const keep = state.hl && state.hl.mod === mod ? state.hl : null;
  const h = state.hl = hlNew(mod);
  if(keep){ h.notes = keep.notes; h.base = keep.base; h.src = keep.src; }
  h.busy = true;
  main.innerHTML = `<div class="empty">Running every check over ${esc(mod)}…<br>
    <span class="count">A few seconds: the map rules are the slow part.</span></div>`;
  let r;
  // the map rules only where the build offers the map screen (2.x ships it off)
  try{ r = await api.get('/api/health?mod=' + enc(mod) + (modeOffered('campmap') ? '&map=1' : '')); }
  catch(e){ r = {error: errText(e)}; }
  if(stale('health', mod) || state.hl !== h) return;
  h.busy = false;
  if(r.error) h.err = r.error; else h.rep = r;
  hlPaint();
}

function hlVisible(f){
  const h = state.hl;
  if(f.severity === 'note' && !h.notes) return false;
  if(f.baseline && h.base) return false;
  if(h.src && f.source !== h.src) return false;
  return true;
}

function hlPaint(){
  const h = state.hl;
  if(h.err){
    main.innerHTML = `<div class="empty">Couldn't run the checks.<br>
      <span class="count">${esc(h.err)}</span><br><br>
      <button class="primary" onclick="loadHealth()">Retry</button></div>`;
    return;
  }
  const r = h.rep, rows = r.findings.filter(hlVisible);
  const labelOf = Object.fromEntries(r.sources.map(s => [s.id, s.label]));
  main.innerHTML = `<div class="hlwrap">
    ${hlHeadHtml(r)}
    ${hlSourcesHtml(r)}
    ${r.when.map(w => hlGroupHtml(w, rows.filter(f => f.when === w.id), labelOf)).join('')}
    ${rows.length ? '' : `<div class="bsec"><div class="trnote">${r.findings.length
      ? 'Nothing matches what is shown. Tick <b>Notes</b> or clear the source above to see the rest.'
      : 'Every check came back clean.'}</div></div>`}
    ${hlSlowHtml(r)}
  </div>`;
}

function hlHeadHtml(r){
  const h = state.hl, c = r.counts;
  const base = r.findings.filter(f => f.baseline).length;
  return `<div class="bsec"><h4>Health of ${esc(r.mod)}
      <span class="count">${r.sources.filter(s => s.state === 'ok').length} checks ran in
        ${(r.ms / 1000).toFixed(1)}s</span>
      <button style="margin-left:auto" onclick="loadHealth()">↻ Run again</button></h4>
    <div class="trnote">Every check the toolkit has, over this mod, in one list. Each row opens the
      screen that owns it, which is where it is fixed. Grouped by when it bites, the way the two
      crash guides in the archive put it.</div>
    <div class="hlcounts">
      <span class="hlc w-bad">✕ ${c.fatal} fatal</span>
      <span class="hlc w-warn">! ${c.warn} warning${c.warn === 1 ? '' : 's'}</span>
      <label class="chk"><input type="checkbox" ${h.notes ? 'checked' : ''}
        onchange="state.hl.notes=this.checked;hlPaint()"> ${c.note} note${c.note === 1 ? '' : 's'}</label>
      ${base ? `<label class="chk" title="The campaign map screen can stamp what a mod already had, so
only what you add afterwards counts. These ${base} were there when it was stamped."><input type="checkbox"
        ${h.base ? 'checked' : ''} onchange="state.hl.base=this.checked;hlPaint()">
        Hide ${base} that were already there</label>` : ''}
    </div></div>`;
}

function hlSourcesHtml(r){
  const h = state.hl;
  return `<div class="hlsrcs">${r.sources.map(s => {
    const n = s.counts.fatal + s.counts.warn + (h.notes ? s.counts.note : 0);
    const cls = s.state === 'failed' ? ' failed' : s.state === 'off' ? ' off' : '';
    const tip = s.state === 'failed' ? `This check could not run: ${s.error}`
      : s.state === 'off' ? 'The campaign map screen is off in this build, so its rules are not run.'
      : `${s.file} · ${s.ms} ms`;
    return `<button class="hlsrc${cls}${h.src === s.id ? ' on' : ''}" title="${esc(tip)}"
      ${s.state === 'ok' ? `onclick="hlSource('${s.id}')"` : 'disabled'}>
      ${esc(s.label)}
      <span class="n">${s.state === 'failed' ? 'failed' : s.state === 'off' ? 'off'
        : s.counts.fatal ? `<b class="w-bad">${s.counts.fatal}</b>${n > s.counts.fatal ? ' +' + (n - s.counts.fatal) : ''}`
        : n}</span></button>`;
  }).join('')}${r.sources.filter(s => s.state === 'failed').map(s =>
    `<div class="hlfail w-bad">✕ ${esc(s.label)} could not run: ${esc(s.error)}</div>`).join('')}</div>`;
}

function hlSource(id){
  const h = state.hl;
  h.src = h.src === id ? '' : id;
  hlPaint();
}

function hlGroupHtml(w, rows, labelOf){
  if(!rows.length) return '';
  const h = state.hl, lim = h.shown[w.id] || HL_PAGE;
  const fatal = rows.filter(f => f.severity === 'fatal').length;
  return `<div class="bsec hlgroup"><h4>${esc(w.label)} <span class="n">${rows.length}</span>
      ${fatal ? `<span class="w-bad">${fatal} fatal</span>` : ''}
      <span class="count">${esc(w.help)}</span></h4>
    <div class="hllist">${rows.slice(0, lim).map(f => hlRowHtml(f, labelOf)).join('')}</div>
    ${rows.length > lim ? `<button class="mini" style="margin-top:6px"
      onclick="state.hl.shown['${w.id}']=${lim + HL_PAGE};hlPaint()">Show ${Math.min(HL_PAGE, rows.length - lim)}
      more of ${rows.length - lim}</button>` : ''}</div>`;
}

function hlRowHtml(f, labelOf){
  const [icon, cls, tip] = HL_SEV[f.severity] || HL_SEV.note;
  const at = f.file ? `${f.file}${f.line ? ':' + f.line : ''}` : '';
  const i = state.hl.rep.findings.indexOf(f);
  return `<div class="hlrow${f.baseline ? ' base' : ''}">
    <span class="hlsev ${cls}" title="${esc(tip)}">${icon}</span>
    <div class="hlmsg">${esc(f.message)}${f.count > 1 ? ` <span class="count">(×${f.count})</span>` : ''}
      <div class="count">${esc(labelOf[f.source] || f.source)}${at ? ` · <code>${esc(at)}</code>` : ''}${
        f.baseline ? ' · already there when the map was stamped' : ''}</div></div>
    <button class="mini" onclick="hlOpen(${i})" title="Open this where it is fixed">Open →</button>
  </div>`;
}

function hlSlowHtml(r){
  return `<div class="bsec"><h4>Cleanup audits <span class="count">not run here</span></h4>
    <div class="trnote">These look for what a mod carries and never uses: unused models, orphan
      files, duplicate cards. None of it stops a mod starting, and each takes from seconds to over a
      minute on a big mod, so each runs on its own screen when you ask.</div>
    <div class="hlslow">${r.slow.map(s => `<div class="hlslowrow">
      <div>${esc(s.label)} <span class="count">${esc(s.cost)}</span></div>
      <button class="mini" onclick="setAppMode('${s.mode}')">Open →</button></div>`).join('')}</div></div>`;
}

/* ---- going to the screen that owns a finding ----
   Each screen loads its own list when it is shown, so the record is opened
   once that list has arrived for THIS mod. `hlWhen` waits for it rather than
   guessing a delay, and gives up quietly after a while: landing on the right
   screen without the record picked is still most of the way there. */
function hlWhen(ready, then, tries = 80){
  let ok = false;
  try{ ok = !!ready(); }catch(e){ ok = false; }
  if(ok){ try{ then(); }catch(e){} return; }
  if(tries > 0) setTimeout(() => hlWhen(ready, then, tries - 1), 150);
}
const hlLoaded = key => () => state[key] && state[key].mod === state.src;

function hlOpen(i){
  const f = state.hl.rep.findings[i];
  if(!f) return;
  const o = f.open || {}, name = o.name || '';
  activity('health', `opened ${f.source} ${f.code}`);
  if(o.mode === 'minor') return minorGo(o.tab || 'rebels'), name &&
    hlWhen(() => state.mf && state.mf.mod === state.src && state.mf.tab === o.tab, () => mfOpen(name));
  if(o.mode === 'edit'){ if(name) return openEditor(name); return setAppMode('edit'); }
  setAppMode(o.mode);
  if(o.mode === 'buildings' && name)
    return hlWhen(() => state.bld && state.bld.mod === state.src && state.bld.ov, () => openBuilding(name));
  if(o.mode === 'campmap') return hlWhen(() => state.cmap && state.cmap.mod === state.src, () => {
    cchkOpen();
    if(!state.cchk.open) cchkToggle();
    if(o.key) hlWhen(() => state.cchk && state.cchk.rep, () => cchkGoKey(o.key));
  });
  if(!name) return;
  const open = {traits: ['tr', trOpen], ancillaries: ['an', anOpen], guilds: ['gu', guOpen],
                factions: ['fac', facOpen], cultures: ['mf', mfOpen]}[o.mode];
  if(open) hlWhen(hlLoaded(open[0]), () => open[1](name));
}
